pepper-apis/src/main/java/org/broadinstitute/ddp/filter/UserAuthCheckFilter.java

```diff
@@ -115,27 +114,33 @@ private void handleTemporaryUser(Request request) {
 
         if (!inWhitelist) {
             LOG.warn("Request '{} {}' is not in temp-user whitelist", requestedMethod, requestedPath);
-            throw halt(401);
+            throw ResponseUtil.haltError(HttpStatus.SC_UNAUTHORIZED,
+                    new ApiError(ErrorCodes.AUTH_CANNOT_BE_DETERMINED, "Request is not in temp-user whitelist"));
         }
 
         UserDto tempUser = TransactionWrapper.withTxn(handle ->
                 handle.attach(JdbiUser.class).findByUserGuid(tempUserGuid));
 
         boolean canAccess = true;
+        String message = null;
         if (tempUser == null) {
             LOG.warn("Could not find temporary user with guid '{}'", tempUserGuid);
+            message = "Could not find temporary user with guid: " + tempUserGuid;
             canAccess = false;
         } else if (!tempUser.isTemporary()) {
             LOG.error("User with guid '{}' is not a temporary user but is used to access"
                     + " temp-user whitelisted path without a token", tempUserGuid);
+            message = "User is not a temporary user";
             canAccess = false;
         } else if (tempUser.isExpired()) {
             LOG.error("Temporary user with guid '{}' had already expired at time {}ms", tempUserGuid, tempUser.getExpiresAtMillis());
+            message = "Temporary user passed expiration time";
             canAccess = false;
         }
 
         if (!canAccess) {
-            throw halt(401);
+            ApiError apiError = new ApiError(ErrorCodes.AUTH_CANNOT_BE_DETERMINED, message);
+            throw ResponseUtil.haltError(HttpStatus.SC_UNAUTHORIZED, apiError);
```

Comment:
I'm still not too sure about why we need to add a body to `401`. Correct me if I'm wrong, but I recall that the point of just returning a `401` is so that we don't leak any information. If we're really going to add a body, I suggest making it as general as possible, e.g.
```
{
  "code": "UNAUTHORIZED",
  "message": "Unauthorized"
}
```

Reply:
Probably should have been discussed while sizing and approving this ticket. 

I have seen three types of patterns in API's:
1. halt(401) no body
2. Return Error (just code) object in Response 
3. Return ApiError (both code & message) object in Response 
From the API tickets in this sprint , looks like we are moving towards # 3

Should we try to keep things consistent OR ok to use all 3 ways depending on the API. (Probably process meeting topic !)
If we want to implement yufengs above suggestions, we probably don't even need to use ApiError object , just return code if we are keeping the message same as code OR again for consistency use ApiError !!
We need to make sure we are not including any info that can lead to information leak/security concern while returning ApiError with message.

Reply:
Have a comment regarding the overarching error structure coming, but in the meantime- should we be using a `401` here? A `403` might be better since a `401` implies that it can be fixed if the client performs the proper authorization.

Reply:
Returning a response body isn't necessarily leaking- in this case, it's just following the pattern for our platform of including error details. I like returning a more vague `ApiError`, and logging more verbose details to the back end, but we still need to include some details so the caller can figure out what's up without access to the server. What if we restrict all client-facing errors in this filter to:

_if the client ID is not stored, enabled, whitelisted, etc- just say it's not authorized._
```json
{
    "code": "UNAUTHORIZED",
    "message": "client $CLIENT_ID_FROM_REQUEST is not authorized"
}
```

_the client ID cleared whatever checks we implemented, but access to the study is restricted for one reason or another (it's disabled, doesn't exist, etc)_
```json
{
    "code": "UNAUTHORIZED",
    "message": "study $STUDY_ID_FROM_REQUEST is not authorized"
}
```

_this could easily be used for temporary users as well, since we have a GUID. No need to treat them any differently in the eyes of the client :) _
```json
{
    "code": "UNAUTHORIZED",
    "message": "user $USER_ID_FROM_REQUEST is not authorized"
}
```

Reply:
@ssettipalli #2 is outdated and is in the process of being removed in preference to #3. 

Regarding the `ApiError` message, what will a client do with that message? Why not just return a generic message in all cases?

Reply:
At least in the sense of the iOS client, those errors tend to be more for the developer and application logging than the end user. The messaging isn't consistent or useful enough to present to the user in a non-confusing manner, so   the client tries its best to take action based on the result, and lets the user know when it has no idea what to do. In this case, presenting a user with a message reading simply "unauthorized" is frustrating and not helpful.

Reply:
Probably returned message should be more meaningful & useful rather than same or same meaning in a sentence as code

Reply:
I had been under impression that we don't want to return any additional information other than a `401` so we don't give attackers any chances. I recall that we mentioned that if we returned some message about user or study, then you can infer that this particular user/study exists.

Can someone check in with the rest of the team about this change?

Reply:
It should, but we still need to define who it's supposed to be useful & meaningful **to**. I feel we should be targeting the developers with most of our errors- popping up an alert to the user should be the worst-case scenario when handling errors.

Reply:
Should I flag this ticket/PR for now and we can discuss during next process meeting !!!

Reply:
The issues arises when there are different responses depending on the user's state. In this case, we also need to balance that with the utility of the message itself. I don't feel that these messages share an unreasonable amount of information- aren't saying whether the user (or client) exists or not, just that whatever GUID was passed in isn't authorized.

If the client sending request only gets part of the request correct, given the length of the identifiers, it's far more likely that it's going to be a developer attempting to get their client working than an attacker.

> I had been under impression that we don't want to return any additional information other than a `401` so we don't give attackers any chances. I recall that we mentioned that if we returned some message about user or study, then you can infer that this particular user/study exists.